grails-app/services/au/org/ala/biocache/hubs/WebServicesService.groovy

```diff
@@ -659,6 +659,32 @@ class WebServicesService {
         dataQualityCodes
     }
 
+    // maps from country name to iso code
+    @Cacheable('longTermCache')
+    def getCountryNameMap() {
+        def countryUrl = "${grailsApplication.config.userdetails.baseUrl}/ws/registration/countries.json"
+        def countries = getJsonElements(countryUrl)
+        return countries?.findAll {it -> beAValidCountryOrState(it as JSONObject)}?.collectEntries { [(String)it.get("name"), (String)it.get("isoCode")] }
+    }
+
+
+    private static boolean beAValidCountryOrState(JSONObject obj) {
+        return obj.has("isoCode") && obj.has("name") && obj.get("isoCode") != "" && obj.get("name") != "N/A"
+    }
+
+    @Cacheable('longTermCache')
+    List<String> getStates(String countryName) {
+        Map countryNameMap = grailsApplication.mainContext.getBean('webServicesService').getCountryNameMap()
+        // if a known country name
+        if (countryNameMap?.containsKey(countryName)) {
+            def states = getJsonElements("${grailsApplication.config.userdetails.baseUrl}/ws/registration/states.json?country=" + countryNameMap.get(countryName))
+            if (states) {
+                // only return valid states
+                return states.findAll { it -> beAValidCountryOrState(it) }.collect { it -> (String)it.get("name") }
+            }
+        }
+        []
```

Comment:
My preference is for a **single return statement**, when it is practical (you have 2 in this method) - as it makes the code easier to read. E.g.

```groovy
List<String> getStates(String countryName) {
    List<String> matchingStates = []

    if (countryNameMap?.containsKey(countryName)) {
        def states = getJsonElements("${grailsApplication.config.userdetails.baseUrl}/ws/registration/states.json?country=" + countryNameMap.get(countryName))
        if (states) {
            // only return valid states
            matchingStates = states.findAll { it -> beAValidCountryOrState(it) }.collect { it -> (String)it.get("name") }
        }
    }

    matchingStates
}
```

Seeing `[]` as last statement is not ideal as it makes me think this is the "intended" output when its really the "exception" to the output.